### PR TITLE
MAINT: more python <3.6  cleanup

### PR DESCRIPTION
**Pull request by bsipocz**
I'm not sure about the second to last commit, but I suppose CI should smoke it out if it was a wrong assumption.

### CODE REVIEW

**Comment by bsipocz** [comment]
OK, so it seems I run into some misunderstanding of the comments, so rather than removing code the comment should be updated to be less specific of python versions

File: numpy/testing/tests/test_utils.py [hunk 1]
```diff
@@ -1240,9 +1237,7 @@ def assert_warn_len_equal(mod, n_in_context, py34=None, py37=None):
         if sys.version_info[:2] >= (3, 7):
             if py37 is not None:
                 n_in_context = py37
-        elif sys.version_info[:2] >= (3, 4):
```
Comment (rgommers) [inline numpy/testing/tests/test_utils.py:1243]: This changes the logic for py36, and results in a test failure on Azure.

**Comment by rgommers** [comment]
Thanks for the cleanup @bsipocz. Overall looks good, but one clear test failure, and the Shippable test failures also look concerning.

**Comment by rgommers** [comment]
Oh never mind the Shippable ones, those are showing up on other PRs as well.

File: numpy/core/_add_newdocs.py [hunk 2]
```diff
@@ -1056,8 +1056,7 @@
             decimal numbers, an operation which is better spelt
             ``frombuffer(string, dtype, count)``. If `string` contains unicode
             text, the binary mode of `fromstring` will first encode it into
-            bytes using either utf-8 (python 3) or the default encoding
-            (python 2), neither of which produce sane results.
+            bytes using either utf-8 which does not produce sane results.
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py:1059]:
```suggestion
            bytes using utf-8 which does not produce sane results.
```

File: numpy/core/_add_newdocs.py [hunk 3]
```diff
@@ -1056,8 +1056,7 @@
             decimal numbers, an operation which is better spelt
             ``frombuffer(string, dtype, count)``. If `string` contains unicode
             text, the binary mode of `fromstring` will first encode it into
-            bytes using either utf-8 (python 3) or the default encoding
-            (python 2), neither of which produce sane results.
```
Comment (eric-wieser) [inline numpy/core/_add_newdocs.py:1060]: For deprecation explanations, I'd argue we should keep the references to old python versions. People upgrading their code want to know what it _used_ to do in their old numpy version, which may still support the older python versions.

File: numpy/core/include/numpy/npy_3kcompat.h [hunk 4]
```diff
@@ -1,6 +1,6 @@
 /*
  * This is a convenience header file providing compatibility utilities
- * for supporting Python 2 and Python 3 in the same code base.
+ * for supporting different minor versions of Python 3.
```
Comment (eric-wieser) [inline numpy/core/include/numpy/npy_3kcompat.h:3]:
```suggestion
 * for supporting different minor versions of Python 3.
 * It was originally used to support the transition from Python 2,
 * hence the "3k" naming.
```

File: numpy/lib/format.py [hunk 5]
```diff
@@ -370,7 +370,7 @@ def _wrap_header(header, version):
     import struct
     assert version is not None
     fmt, encoding = _header_size_info[version]
-    if not isinstance(header, bytes):  # always true on python 3
```
Comment (eric-wieser) [inline numpy/lib/format.py:373]: Was this comment wrong? If it wasn't, we should unfold / remove this `if`.

**Comment by mattip** [comment]
@bsipocz ping

**Comment by charris** [comment]
@bsipocz Do you have time to finish this up?

**Comment by bsipocz** [comment]
I'm sorry it took a while. This is now rebased, and the comments from the first review is addressed or became obsolete since then. I'll see if CI picks up any issues, if not this should be ready.
I suppose some more cleanup can be done for python 3.6 and 3.7, but I would rather do that in a follow-up than drag this around much longer.

File: numpy/core/include/numpy/npy_3kcompat.h (hunk 4 above)
Comment (mattip) [reply numpy/core/include/numpy/npy_3kcompat.h:3]: Did you want to commit this?

**Review by mattip — APPROVED** [review]
LGTM pending CI success. I restarted the macOS build that timed out.

**Comment by mattip** [comment]
Thanks @bsipocz

**Pull request merged**